### ISSUE TITLE
Visualization

### ISSUE THREAD

**Issue by Osos8128** [jonathanking/sidechainnet#15]
I'm working on colab and i see your code in colab walkthrough 
       **s = build_visualizable_structures(seqonly_model, d["train"], mode="seqs")**
i see that you are in this line function return a batch of 3d shapes so when you call **s.to_3Dmol()** which protein are visualized because i think **s** should contains more than one protein. also what is idx that to_3Dmol() expect ??!

**Comment by jonathanking** [comment]
`build_visualizable_structures` is a function I built for demonstration purposes only in the colab notebook. It returns a `BatchedStructureBuilder` which allows you to visualize the structures for a batch of proteins. 

Perhaps the confusion is that I missed a `break` statement in the colab. My intent was to return a `BatchedStructureBuilder` for a single batch, but I mistakenly looped through the training set. I have updated the function now.

`to_3Dmol()` requests an index that corresponds to the index in the batch dimension, i.e. the 1st, 2nd, etc. protein in a `BatchedStructureBuilder`.

Please let me know if I can clarify any further!

**Comment by Osos8128** [comment]
My function had this break and i only return one batch but when i write any idx it gives me this error
only idx 1 work.
this a screenshot from my work you will find my code is similar to yours.
![TempIssueJon](https://user-images.githubusercontent.com/40743772/107432364-8641e600-6b30-11eb-8e87-c39f90be00c3.png)

**Comment by jonathanking** [comment]
Can you share with me your call to `scn.load`? I'm not sure that I've encountered this error before.

**Comment by Osos8128** [comment]
That's it.
![TempP2Jon](https://user-images.githubusercontent.com/40743772/107435429-cf943480-6b34-11eb-881b-278f2a9dfd8a.png)

**Comment by Osos8128** [comment]
That's it
![TempP2Jon](https://user-images.githubusercontent.com/40743772/107435619-0f5b1c00-6b35-11eb-9437-9eb84fe8545f.png)

**Comment by jonathanking** [comment]
Hmm, I see. Are you willing to share your notebook with me (jekshmeck@gmail.com)? I would be able to take a look. If not, I would appreciate it if you could expand the error traceback where it says "6 Frames" so I can see what might have happened.

**Comment by Osos8128** [comment]
I shared it with you.

**Comment by jonathanking** [comment]
Thanks! That was really helpful.

I was able to figure out the issue. Your structure had some angles with a value of 0. When the StructureBuilder tried to build it using those angles, I had an if statement that was checking if the value of some variable was `None`, but it mistakenly caused an error because the value was 0. It should work now without this error.

**Comment by Osos8128** [comment]
Thanks for your time.

**Comment by Osos8128** [comment]
Thanks for your time.
I'm facing another problem so if you have time to look at it because my notebook already shared with you i will be grateful. if you don't have time don't bother yourself.
my model now should be done and training but i think that my loss calculation is wrong somewhere because of loss values that comes while training. so if you could review my drmsd calculations if it is correct or not.

**Comment by jonathanking** [comment]
You're welcome!

I took a look, but it's hard to tell. Unfortunately, there are a lot of things that might be going wrong. If you are interested, I have built several DRMSD related functions available in examples/losses. Hopefully those can help!

I have a question for you, if you don't mind. When you define the rotation matrix D2, you use `-R` for the first entry (`-R cos(T)`). Can you explain why the negative is there? In the formulation provided in Mohammad AlQuraishi's paper, there is no negative sign. I have observed that I need the negative sign in my work too, but I just am looking to remind myself why it is needed here and why there is no negative sign in some of the written equations.

**Comment by Osos8128** [comment]
You mean a lot of things in the model not only in dRMSD ??!! :(

i tried to use your dRMSD function and change NUM_CRDS_PER_RES to 3 but i got shape error.
so i think that i have to replace pred_crds with true_crds in batch then calculate loss using your function.
or pred_angs with true_angs then build it using your build function then calc loss.
but i think if i do that loss will not be accurate due to sidechain coords and also because of you have 4 atoms per backbone residue not 3 like qurashi. it is hard and complicated for me to merge your work into qurashi model.

i wish i could tell you but i just understand the equation from paper and copy the function from another work that also trying to reporduce qurashi work in pytorch but his model is not completed yet. https://github.com/conradry/pytorch-rgn
you could ask him im sure he will have an clear answer for your question.

**Comment by jonathanking** [comment]
I didn't mean to say that something is wrong for certain, but, unfortunately, I don't have time at the moment to debug the DRMSD function you have written. I can understand that modifying the code to predict three atoms only may be challenging. It's not something that I have worked on myself, but if I ever implement some of this in the future, I will let you know. I'm sure it will be useful to several people.

Thanks for clarifying about the negative sign!

**Comment by Osos8128** [comment]
I got it. I am going to review all of my code again.
Thanks again for your help and for this great library and i will let you know if my work produce good results.
